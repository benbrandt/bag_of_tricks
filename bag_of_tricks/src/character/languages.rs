use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use super::Character;

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Auran,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub(crate) fn gen(rng: &mut impl Rng, character: &Character, amount: usize) -> Vec<Self> {
        Language::iter()
            .filter(|l| !character.languages.contains(l))
            .choose_multiple(rng, amount)
    }
}

/// Trait for describing what languages are provided by a given object.
pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> usize {
        0
    }
}
