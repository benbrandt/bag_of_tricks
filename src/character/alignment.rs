use std::fmt;

use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Morality {
    Evil,
    Good,
    Neutral,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Alignment(Attitude, Morality);

impl Alignment {
    pub(crate) fn gen(rng: &mut impl Rng) -> Self {
        let attitude = Attitude::iter().choose(rng).unwrap();
        let morality = Morality::iter().choose(rng).unwrap();
        Self(attitude, morality)
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
