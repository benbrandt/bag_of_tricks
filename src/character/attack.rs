use std::fmt;

use strum_macros::Display;

use crate::dice_roller::RollCmd;

#[allow(dead_code)]
#[derive(Display)]
pub(crate) enum DamageType {
    Acid,
    Bludgeoning,
    Cold,
    Fire,
    Force,
    Lightning,
    Necrotic,
    Piercing,
    Poison,
    Psychic,
    Radiant,
    Slashing,
    Thunder,
}

pub(crate) struct Damage {
    pub(crate) damage_type: DamageType,
    pub(crate) modifier: i16,
    pub(crate) roll: RollCmd,
}

impl fmt::Display for Damage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}{:+} {}", self.roll, self.modifier, self.damage_type)
    }
}

// pub(crate) enum Hit {
//     Roll(u8),
//     DC(AbilityScoreType, i16),
// }

// impl fmt::Display for Hit {
//     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//         match self {
//             Self::Roll(m) => write!(f, "{:+}", m),
//             Self::DC(a, s) => write!(f, "{} {}", s, a),
//         }
//     }
// }

// pub(crate) struct Attack<'a> {
//     pub(crate) citation: Citation,
//     pub(crate) hit: Hit,
//     pub(crate) damage: Damage,
//     pub(crate) name: &'a str,
//     pub(crate) range: &'a str,
// }

// impl fmt::Display for Attack<'_> {
//     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//         writeln!(
//             f,
//             "{:20} {:20} {:20} {:20} {}",
//             self.name, self.range, self.hit, self.damage, self.citation
//         )
//     }
// }
