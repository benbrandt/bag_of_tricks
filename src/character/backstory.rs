pub(crate) trait Backstory {
    fn backstory(&self) -> Vec<String> {
        vec![]
    }
}
