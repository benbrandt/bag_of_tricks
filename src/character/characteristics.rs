use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use std::{fmt, ops::RangeInclusive};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use crate::dice_roller::RollCmd;

/// An range of ages a given adventurer could be.
pub(crate) struct AgeRange(pub(crate) RangeInclusive<u16>);

impl AgeRange {
    /// Generate a random age within the given range.
    fn gen(&self, rng: &mut impl Rng) -> u16 {
        rng.gen_range(self.0.clone())
    }
}

/// Really only here to help decide on names, not core to a character choice.
#[derive(Deserialize, Display, EnumIter, Serialize)]
pub(crate) enum Gender {
    Female,
    Male,
}

impl Gender {
    /// Choose a random gender
    fn gen(rng: &mut impl Rng) -> Self {
        Gender::iter().choose(rng).unwrap()
    }
}

/// Weight modifier
#[derive(Debug, PartialEq)]
pub(crate) enum WeightMod {
    /// Fixed additional weight
    Fixed(u16),
    /// Role to find additional weight
    Roll(RollCmd),
}

/// Convert feet + inches to just inches (allow for better readability in tables)
pub(crate) const fn in_inches(feet: u8, inches: u8) -> u8 {
    feet * 12 + inches
}

/// Data from a table to generate a height and weight for a given character.
#[derive(Debug, PartialEq)]
pub(crate) struct HeightAndWeightTable {
    /// Base height for this character
    pub(crate) base_height: u8,
    /// Base weight for this character
    pub(crate) base_weight: u16,
    /// Modifier to apply to the height
    pub(crate) height_mod: RollCmd,
    /// Modifier to apply to the weight (informed by height)
    pub(crate) weight_mod: WeightMod,
}

impl HeightAndWeightTable {
    /// Generate a random height and weight for a character
    pub(crate) fn gen(&self, rng: &mut impl Rng) -> (usize, usize) {
        let h = self.height_mod.roll(rng).total();
        // Weight modifier is multiplied by height
        let w = h * match &self.weight_mod {
            WeightMod::Fixed(f) => usize::from(*f),
            WeightMod::Roll(r) => r.roll(rng).total(),
        };
        (
            usize::from(self.base_height) + h,
            usize::from(self.base_weight) + w,
        )
    }
}

/// Size of character (there are more options for monsters)
#[derive(Deserialize, Display, EnumIter, Serialize)]
pub(crate) enum Size {
    Small,
    Medium,
}

/// Types of movement speeds
#[derive(Clone, Copy, Debug, Deserialize, EnumIter, PartialEq, Serialize)]
pub(crate) enum Speed {
    Climbing(u8),
    Swimming(u8),
    Walking(u8),
}

impl fmt::Display for Speed {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Climbing(s) => write!(f, "Climbing Speed: {}ft", s),
            Self::Swimming(s) => write!(f, "Swimming Speed: {}ft", s),
            Self::Walking(s) => write!(f, "Walking Speed: {}ft", s),
        }
    }
}

/// Physical characteristics about a character.
#[derive(Deserialize, Serialize)]
pub(crate) struct CharacteristicDetails {
    /// Age of the character
    pub(crate) age: u16,
    /// Base speed of the character
    pub(crate) base_speeds: Vec<Speed>,
    /// Gender of the character (only used for name choices)
    pub(crate) gender: Gender,
    /// Height of the character
    pub(crate) height: usize,
    /// Size of the character
    pub(crate) size: Size,
    /// Weight of the character
    pub(crate) weight: usize,
}

impl fmt::Display for CharacteristicDetails {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Age: {}", self.age)?;
        writeln!(f, "Gender: {}", self.gender)?;
        writeln!(f, "Size: {}", self.size)?;
        writeln!(f, "Height: {}'{}\"", self.height / 12, self.height % 12)?;
        writeln!(f, "Weight: {} lb.", self.weight)
    }
}

/// Trait for object (usually race) to generate appropriate characteristics about a character
///
/// Separate from race since it uses associated constants (which can't be on a trait object)
pub(crate) trait Characteristics {
    /// Age range to choose from
    const AGE_RANGE: AgeRange;
    /// Character size
    const SIZE: Size;

    /// Base speeds
    fn get_base_speeds(&self) -> Vec<Speed>;

    /// Calculate the height and weight table to generate from
    fn get_height_and_weight_table(&self) -> &HeightAndWeightTable;

    /// Generate characteristics for this race
    fn gen_characteristics(&self, rng: &mut impl Rng) -> CharacteristicDetails {
        let (height, weight) = self.get_height_and_weight_table().gen(rng);
        CharacteristicDetails {
            age: Self::AGE_RANGE.gen(rng),
            base_speeds: self.get_base_speeds(),
            gender: Gender::gen(rng),
            height,
            size: Self::SIZE,
            weight,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_age_range() {
        let mut rng = Pcg64::from_entropy();
        let age = AgeRange(1..=100).gen(&mut rng);
        assert!((1..=100).contains(&age));
    }

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_characteristics() {
        let mut rng = Pcg64::seed_from_u64(1);
        let characteristics = CharacteristicDetails {
            age: 100,
            base_speeds: vec![Speed::Walking(30)],
            gender: Gender::gen(&mut rng),
            height: 75,
            size: Size::Medium,
            weight: 300,
        };
        insta::assert_yaml_snapshot!(characteristics);
    }

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_characteristics_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let characteristics = CharacteristicDetails {
            age: 100,
            base_speeds: vec![Speed::Walking(30)],
            gender: Gender::gen(&mut rng),
            height: 75,
            size: Size::Medium,
            weight: 300,
        };
        insta::assert_snapshot!(format!("{}", characteristics));
    }
}
