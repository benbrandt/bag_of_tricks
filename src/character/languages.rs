use serde::{Deserialize, Serialize};
use strum_macros::{Display, EnumIter};

#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Does this provide additional language options?
    fn addl_languages(&self) -> usize {
        0
    }
}
