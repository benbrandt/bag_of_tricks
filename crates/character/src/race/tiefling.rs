#![allow(clippy::default_trait_access)]
use std::fmt;

use alignment::{AlignmentInfluences, Attitude, Morality};
use attack::{DamageType, Resistances};
use backstory::Backstory;
use characteristics::{
    in_inches,
    names::{
        tiefling::{FEMALE_ABYSSAL, MALE_ABYSSAL, SURNAMES, VIRTUE_NAMES},
        Name,
    },
    AgeRange, Appearance, CharacteristicDetails, Characteristics, Gender, HeightAndWeightTable,
    Size, Speed, WeightMod,
};
use citation::{Book, Citation, CitationList, Citations};
use dice_roller::{Die, RollCmd};
use languages::{Language, Languages};
use personality::PersonalityOptions;
use rand::{
    prelude::{IteratorRandom, SliceRandom},
    Rng,
};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};
use trinkets::Trinkets;

use crate::{
    ability::{AbilityScore, AbilityScoreType},
    features::{Feature, Features},
    proficiencies::Proficiencies,
};

use super::{human::Human, Race};

/// Tiefling height and weight table
const HEIGHT_AND_WEIGHT: HeightAndWeightTable = HeightAndWeightTable {
    base_height: in_inches(4, 9),
    base_weight: 110,
    height_mod: RollCmd(2, Die::D8),
    weight_mod: WeightMod::Roll(RollCmd(2, Die::D4)),
};

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum HornShape {
    Antelope,
    Gazelle,
    Ram,
}

impl Default for HornShape {
    fn default() -> Self {
        Self::Antelope
    }
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum HornSize {
    Large,
    Small,
}

impl Default for HornSize {
    fn default() -> Self {
        Self::Large
    }
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum SkinColor {
    #[strum(serialize = "Dark Blue")]
    DarkBlue,
    Red,
}

impl Default for SkinColor {
    fn default() -> Self {
        Self::DarkBlue
    }
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum SkinTexture {
    Leathery,
    Scaly,
}

impl Default for SkinTexture {
    fn default() -> Self {
        Self::Leathery
    }
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Teeth {
    Fangs,
    #[strum(serialize = "sharp teeth")]
    Sharp,
}

impl Default for Teeth {
    fn default() -> Self {
        Self::Fangs
    }
}

#[derive(Deserialize, EnumIter, Serialize)]
enum PhysicalAppearance {
    Brimstone,
    CatlikeEyes,
    ClovenHooves,
    ForkedTail,
    ForkedTongue,
    GoatLegs,
    Horns(HornSize, HornShape),
    NoShadow,
    SkinColor(SkinColor),
    SkinTexture(SkinTexture),
    SixFingers,
    Teeth(Teeth),
}

impl PhysicalAppearance {
    fn gen(rng: &mut impl Rng) -> Vec<Self> {
        let amount = RollCmd(1, Die::D4).roll(rng).total() + 1;
        PhysicalAppearance::iter()
            .choose_multiple(rng, amount)
            .into_iter()
            .map(|a| match a {
                Self::Brimstone
                | Self::CatlikeEyes
                | Self::ClovenHooves
                | Self::ForkedTail
                | Self::ForkedTongue
                | Self::GoatLegs
                | Self::NoShadow
                | Self::SixFingers => a,
                Self::Horns(_, _) => Self::Horns(
                    HornSize::iter().choose(rng).unwrap(),
                    HornShape::iter().choose(rng).unwrap(),
                ),
                Self::SkinColor(_) => Self::SkinColor(SkinColor::iter().choose(rng).unwrap()),
                Self::SkinTexture(_) => Self::SkinTexture(SkinTexture::iter().choose(rng).unwrap()),
                Self::Teeth(_) => Self::Teeth(Teeth::iter().choose(rng).unwrap()),
            })
            .collect()
    }
}

impl fmt::Display for PhysicalAppearance {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Brimstone => "exude a smell of brimstone".to_string(),
                Self::CatlikeEyes => "catlike eyes".to_string(),
                Self::ClovenHooves => "cloven hooves".to_string(),
                Self::ForkedTail => "a forked tail".to_string(),
                Self::ForkedTongue => "a forked tongue".to_string(),
                Self::GoatLegs => "goatlike legs".to_string(),
                Self::Horns(size, shape) => format!("{} {} horns", size, shape),
                Self::NoShadow => "cast no shadow or reflection".to_string(),
                Self::SkinColor(c) => format!("{}", c),
                Self::SkinTexture(t) => format!("{}", t),
                Self::SixFingers => "six fingers on each hand".to_string(),
                Self::Teeth(t) => format!("{}", t),
            }
        )
    }
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum FeralVariant {
    #[strum(serialize = "Devil's Tongue")]
    DevilsTongue,
    Hellfire,
    Winged,
}

impl FeralVariant {
    fn features(&self) -> Vec<Feature> {
        match self {
            // You know the Vicious Mockery cantrip. Once you reach 3rd level, you can cast the Charm Person spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Enthrall spell once. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells. This trait replaces the Infernal Legacy trait
            Self::DevilsTongue => vec![Feature {
                title: "Devil's Tongue",
                citation: Citation(Book::Scag, 118),
            }],
            // Once you reach 3rd level, you can cast the Burning Hands spell once as a 2nd-level spell. This trait replaces the Hellish Rebuke spell of the Infernal Legacy trait.
            Self::Hellfire => vec![Feature {
                title: "Hellfire",
                citation: Citation(Book::Scag, 118),
            }],
            Self::Winged => vec![],
        }
    }
}

impl Default for FeralVariant {
    fn default() -> Self {
        Self::DevilsTongue
    }
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum TieflingSubrace {
    Asmodeus,
    Baalzebul,
    Dispater,
    Feral(FeralVariant),
    Fierna,
    Glasya,
    Levistus,
    Mammon,
    Mephistopheles,
    Zariel,
}

impl TieflingSubrace {
    fn gen(rng: &mut impl Rng) -> Self {
        let subrace = Self::iter().choose(rng).unwrap();
        if matches!(subrace, Self::Feral(_)) {
            Self::Feral(FeralVariant::iter().choose(rng).unwrap())
        } else {
            subrace
        }
    }
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Tiefling {
    appearance: Vec<PhysicalAppearance>,
    subrace: TieflingSubrace,
}

impl Tiefling {
    /// Generate first name for a tiefling. Could be abyssal, human, or virtue name.
    fn gen_first_name(rng: &mut impl Rng, characteristics: &CharacteristicDetails) -> &'static str {
        let abyssal_name = *(match characteristics.gender {
            Gender::Female => FEMALE_ABYSSAL,
            Gender::Male => MALE_ABYSSAL,
        })
        .choose(rng)
        .unwrap();
        let human_name = Human::gen_first_name(rng, characteristics);
        let virtue_name = *VIRTUE_NAMES.choose(rng).unwrap();
        *[abyssal_name, human_name, virtue_name].choose(rng).unwrap()
    }

    // Generate surname, could be human, or more traditional tiefling
    fn gen_surname(rng: &mut impl Rng, characteristics: &CharacteristicDetails) -> &'static str {
        *[
            Human::gen_surname(rng, characteristics),
            SURNAMES.choose(rng).unwrap(),
        ]
        .choose(rng)
        .unwrap()
    }
}

impl AlignmentInfluences for Tiefling {
    fn attitude(&self) -> Vec<Attitude> {
        vec![Attitude::Chaotic]
    }

    fn morality(&self) -> Vec<Morality> {
        vec![Morality::Evil]
    }
}

impl Appearance for Tiefling {
    fn appearance(&self) -> Vec<String> {
        let mut appearance: Vec<String> =
            self.appearance.iter().map(|a| format!("{}", a)).collect();
        if matches!(self.subrace, TieflingSubrace::Feral(FeralVariant::Winged)) {
            appearance.push("Bat-like wings jut from your shoulder blades. You have a flying speed of 30ft while not wearing heavy armor".to_string());
        }
        appearance
    }
}

impl Backstory for Tiefling {}

impl Characteristics for Tiefling {
    const HUMAN_ANCESTRY: bool = true;
    const SIZE: Size = Size::Medium;

    fn get_age_range(&self) -> AgeRange {
        AgeRange(10..=100)
    }

    fn get_base_speeds(&self) -> Vec<Speed> {
        let mut speeds = vec![Speed::Walking(30)];
        if matches!(self.subrace, TieflingSubrace::Feral(FeralVariant::Winged)) {
            speeds.push(Speed::Flying(30));
        }
        speeds
    }

    fn get_height_and_weight_table(&self) -> &HeightAndWeightTable {
        &HEIGHT_AND_WEIGHT
    }
}

impl Citations for Tiefling {
    fn citations(&self) -> CitationList {
        let mut citations = vec![Citation(Book::Phb, 42)];
        citations.push(match self.subrace {
            TieflingSubrace::Asmodeus
            | TieflingSubrace::Baalzebul
            | TieflingSubrace::Dispater
            | TieflingSubrace::Fierna => Citation(Book::Mtof, 21),
            TieflingSubrace::Feral(_) => Citation(Book::Scag, 118),
            TieflingSubrace::Glasya | TieflingSubrace::Levistus | TieflingSubrace::Mammon => {
                Citation(Book::Mtof, 22)
            }
            TieflingSubrace::Mephistopheles | TieflingSubrace::Zariel => Citation(Book::Mtof, 23),
        });
        CitationList(citations)
    }
}

impl Features for Tiefling {
    fn features(&self) -> Vec<Feature> {
        let mut features = vec![
            // Thanks to your infernal heritage, you have superior vision in dark and dim conditions. You can see in dim light within 60 feet of you as if it were bright light, and in darkness as if it were dim light. You can't discern color in darkness, only shades of gray.
            Feature {
                title: "Darkvision",
                citation: Citation(Book::Phb, 43),
            },
        ];
        features.extend(match &self.subrace {
            // You know the thaumaturgy cantrip. When you reach 3rd level, you can cast the hellish rebuke spell as a 2nd-level spell once with this trait and regain the ability to do so when you finish a long rest. When you reach 5th level, you can cast the darkness spell once with this trait and regain the ability to do so when you finish a long rest. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Asmodeus => {
                vec![Feature {
                    title: "Infernal Legacy",
                    citation: Citation(Book::Phb, 43),
                }]
            }
            // You know the Thaumaturgy cantrip. Once you reach 3rd level, you can cast the Ray of Sickness spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Crown of Madness spell once. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Baalzebul => vec![Feature {
                title: "Legacy of Maladomini",
                citation: Citation(Book::Mtof, 21),
            }],
            // You know the Thaumaturgy cantrip. Once you reach 3rd level, you can cast the Disguise Self spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Detect Thoughts spell once. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Dispater => vec![Feature {
                title: "Legacy of Dis",
                citation: Citation(Book::Mtof, 21),
            }],
            TieflingSubrace::Feral(v) => v.features(),
            // You know the Friends cantrip. Once you reach 3rd level, you can cast the Charm Person spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Suggestion spell once. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Fierna => vec![Feature {
                title: "Legacy of Phlegethos",
                citation: Citation(Book::Mtof, 21),
            }],
            // You know the Minor Illusion cantrip. Once you reach 3rd level, you can cast the Disguise Self spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Invisibility spell once as a 2nd-level spell. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Glasya => vec![Feature {
                title: "Legacy of Malbolge",
                citation: Citation(Book::Mtof, 22),
            }],
            // You know the Ray of Frost cantrip. Once you reach 3rd level, you can cast the Armor of Agathys spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Darkness spell once. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Levistus => vec![Feature {
                title: "Legacy of Stygia",
                citation: Citation(Book::Mtof, 22),
            }],
            // You know the Mage Hand cantrip. Once you reach 3rd level, you can cast the Tenser's Floating Disk spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Arcane Lock spell once. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Mammon => vec![Feature {
                title: "Legacy of Minauros",
                citation: Citation(Book::Mtof, 22),
            }],
            // You know the Mage Hand cantrip. Once you reach 3rd level, you can cast the Burning Hands spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Flame Blade spell once as a 3rd-level spell. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Mephistopheles => vec![Feature {
                title: "Legacy of Cania",
                citation: Citation(Book::Mtof, 23),
            }],
            // You know the Thaumaturgy cantrip. Once you reach 3rd level, you can cast the Searing Smite spell once as a 2nd-level spell. Once you reach 5th level, you can also cast the Branding Smite spell once as a 3rd-level spell. You must finish a long rest to cast these spells again with this trait. Charisma is your spellcasting ability for these spells.
            TieflingSubrace::Zariel => vec![Feature {
                title: "Legacy of Avernus",
                citation: Citation(Book::Mtof, 23),
            }],
        });

        features
    }
}

impl Languages for Tiefling {
    /// Names given by tiefling race
    fn languages(&self) -> Vec<Language> {
        vec![Language::Common, Language::Infernal]
    }
}

impl Name for Tiefling {
    /// Name also requires getting a set of human names (for human lineage)
    fn gen_name(rng: &mut impl Rng, characteristics: &CharacteristicDetails) -> String {
        format!(
            "{} {}",
            Self::gen_first_name(rng, characteristics),
            Self::gen_surname(rng, characteristics)
        )
    }
}

impl PersonalityOptions for Tiefling {}

impl Proficiencies for Tiefling {}

#[typetag::serde]
impl Race for Tiefling {
    fn gen(rng: &mut impl Rng) -> (Box<dyn Race>, String, CharacteristicDetails) {
        let race = Box::new(Self {
            appearance: PhysicalAppearance::gen(rng),
            subrace: TieflingSubrace::gen(rng),
        });
        let characteristics = race.gen_characteristics(rng);
        let name = Self::gen_name(rng, &characteristics);
        (race, name, characteristics)
    }

    fn abilities(&self) -> Vec<AbilityScore> {
        match self.subrace {
            TieflingSubrace::Asmodeus
            | TieflingSubrace::Baalzebul
            | TieflingSubrace::Mammon
            | TieflingSubrace::Mephistopheles => {
                vec![
                    AbilityScore(AbilityScoreType::Charisma, 2),
                    AbilityScore(AbilityScoreType::Intelligence, 1),
                ]
            }
            TieflingSubrace::Dispater | TieflingSubrace::Glasya => vec![
                AbilityScore(AbilityScoreType::Charisma, 2),
                AbilityScore(AbilityScoreType::Dexterity, 1),
            ],
            TieflingSubrace::Feral(_) => vec![
                AbilityScore(AbilityScoreType::Dexterity, 2),
                AbilityScore(AbilityScoreType::Intelligence, 1),
            ],
            TieflingSubrace::Fierna => vec![
                AbilityScore(AbilityScoreType::Charisma, 2),
                AbilityScore(AbilityScoreType::Wisdom, 1),
            ],
            TieflingSubrace::Levistus => vec![
                AbilityScore(AbilityScoreType::Charisma, 2),
                AbilityScore(AbilityScoreType::Constitution, 1),
            ],
            TieflingSubrace::Zariel => vec![
                AbilityScore(AbilityScoreType::Charisma, 2),
                AbilityScore(AbilityScoreType::Strength, 1),
            ],
        }
    }
}

impl Resistances for Tiefling {
    fn resistances(&self) -> Vec<DamageType> {
        vec![DamageType::Fire]
    }
}

impl Trinkets for Tiefling {}

impl fmt::Display for Tiefling {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} Tiefling", self.subrace)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use characteristics::Gender;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let tiefling = Tiefling::gen(&mut rng);
        insta::assert_yaml_snapshot!(tiefling);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let (tiefling, _name, _characteristics) = Tiefling::gen(&mut rng);
        insta::assert_display_snapshot!(tiefling);
    }

    #[test]
    fn test_attitude() {
        let mut rng = Pcg64::seed_from_u64(1);
        let (tiefling, _name, _characteristics) = Tiefling::gen(&mut rng);
        insta::assert_yaml_snapshot!(tiefling.attitude());
    }

    #[test]
    fn test_morality() {
        let mut rng = Pcg64::seed_from_u64(1);
        let (tiefling, _name, _characteristics) = Tiefling::gen(&mut rng);
        insta::assert_yaml_snapshot!(tiefling.morality());
    }

    #[test]
    fn test_appearance() {
        let mut rng = Pcg64::seed_from_u64(1);
        let (tiefling, _name, _characteristics) = Tiefling::gen(&mut rng);
        insta::assert_yaml_snapshot!(tiefling.appearance());
    }

    #[test]
    fn test_snapshot_citations() {
        insta::assert_yaml_snapshot!(TieflingSubrace::iter()
            .map(|subrace| Tiefling {
                appearance: vec![],
                subrace
            }
            .citations())
            .collect::<Vec<CitationList>>());
    }

    #[test]
    fn test_snapshot_features() {
        insta::assert_yaml_snapshot!(TieflingSubrace::iter()
            .map(|subrace| Tiefling {
                appearance: vec![],
                subrace
            }
            .features())
            .collect::<Vec<Vec<Feature>>>())
    }

    #[test]
    fn test_snapshot_languages() {
        let tiefling = Tiefling {
            appearance: vec![],
            subrace: TieflingSubrace::Asmodeus,
        };
        insta::assert_yaml_snapshot!(tiefling.languages());
    }

    #[test]
    fn test_name() {
        let mut rng = Pcg64::seed_from_u64(1);
        let tiefling = Tiefling {
            appearance: vec![],
            subrace: TieflingSubrace::Asmodeus,
        };
        let characteristics_1 = tiefling.gen_characteristics(&mut rng);
        let characteristics_2 = tiefling.gen_characteristics(&mut rng);
        let female_name = Tiefling::gen_name(
            &mut rng,
            &CharacteristicDetails {
                gender: Gender::Female,
                ..characteristics_1
            },
        );
        let male_name = Tiefling::gen_name(
            &mut rng,
            &CharacteristicDetails {
                gender: Gender::Male,
                ..characteristics_2
            },
        );
        insta::assert_yaml_snapshot!([female_name, male_name]);
    }

    #[test]
    fn test_snapshot_abilities() {
        insta::assert_yaml_snapshot!(TieflingSubrace::iter()
            .map(|subrace| Tiefling {
                appearance: vec![],
                subrace
            }
            .abilities())
            .collect::<Vec<Vec<AbilityScore>>>());
    }

    #[test]
    fn test_resistances() {
        let tiefling = Tiefling {
            appearance: vec![],
            subrace: TieflingSubrace::Asmodeus,
        };
        insta::assert_yaml_snapshot!(tiefling.resistances());
    }
}
