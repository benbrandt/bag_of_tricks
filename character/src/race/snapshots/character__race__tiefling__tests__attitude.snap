---
source: character/src/race/tiefling.rs
expression: tiefling.attitude()

---
- Chaotic

